### PR TITLE
ASoC: SOF: create audio buffer page tables in a hardware-safe way

### PR DESCRIPTION
**Pull request by lyakh**
The SOF packs 20-bit page frame numbers into memory back-to-back,
which produces both unaligned access and endianness problems. Use
unaligned memory accessors to guarantee correct packing on all
platforms.

A discussion is also underway to verify whether such packing is at all needed or can be safely dropped for simplicity and performance.

### CODE REVIEW

**Comment by lyakh** [comment]
The CI is complaining about the 80 characters limit in a comment line, which - yes - is 81 or 82 characters long and IMHO has a good reason for that.

File: sound/soc/sof/core.c [hunk 1]
```diff
@@ -215,16 +216,31 @@ int snd_sof_create_page_table(struct snd_sof_dev *sdev,
...
 		u32 pfn = snd_sgbuf_get_addr(dmab, i * PAGE_SIZE) >> PAGE_SHIFT;
-		u32 *pg_table;
+		u8 *pg_table;
 
 		dev_vdbg(sdev->dev, "pfn i %i idx %d pfn %x\n", i, idx, pfn);
 
-		pg_table = (u32 *)(page_table + idx);
+		pg_table = (u8 *)(page_table + idx);
 
+		/*
```
Comment (dbaluta) [inline sound/soc/sof/core.c:225]: The commit is super nice! One thing if you can do is: explain why the compression is needed for readers who will try to understand the code later.

**Comment by plbossart** [comment]
> The CI is complaining about the 80 characters limit in a comment line, which - yes - is 81 or 82 characters long and IMHO has a good reason for that.

don't worry about this, I check where the errors happen and if it's in comments or dev_err logs I ignore them.

**Comment by plbossart** [comment]
I think we should mention that this is an optimization to handle large ring buffers even if the mailboxes are small, but that it's not required in most cases and adds a little-endian restriction so we might make this optional in a future update for cases where this compression is not required. This would be an ABI change though so the compression support will have to remain in the driver for backwards compatibility.

**Pull request merged**